src/main/java/net/syscon/elite/web/config/AuthAwareAuthenticationConverter.java

```diff
@@ -1,43 +1,38 @@
-import org.springframework.util.StringUtils;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter;
 
 import java.util.Collection;
-import java.util.Collections;
+import java.util.HashSet;
 import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
 
-@Component
-public class AuthAwareAuthenticationConverter extends DefaultUserAuthenticationConverter {
-    @Override
-    public Authentication extractAuthentication(final Map<String, ?> map) {
-        if (map.containsKey(USERNAME)) {
-            final var principal = map.get(USERNAME);
-            final var authorities = getAuthorities(map);
-            final var authSourceObject = map.get("auth_source");
-            final var authSource = authSourceObject instanceof String ? (String) authSourceObject : null;
-            return new AuthAwareAuthenticationToken(principal, "N/A", authSource, authorities);
-        }
-        return null;
+public class AuthAwareAuthenticationConverter implements Converter<Jwt, AbstractAuthenticationToken> {
+
+    private final Converter<Jwt, Collection<GrantedAuthority>> jwtGrantedAuthoritiesConverter = new JwtGrantedAuthoritiesConverter();
+
+    public AuthAwareAuthenticationToken convert(final Jwt jwt) {
+        Map<String, Object> map = jwt.getClaims();
+
+        final var principal = map.get("user_name");
+        final var authorities = extractAuthorities(jwt);
+        final var authSourceObject = map.get("auth_source");
+        final var authSource = authSourceObject instanceof String ? (String) authSourceObject : null;
+        return new AuthAwareAuthenticationToken(jwt, principal, authSource, authorities);
     }
 
-    // Copied from DefaultUserAuthenticationConverter as private method
-    private Collection<? extends GrantedAuthority> getAuthorities(final Map<String, ?> map) {
-        if (!map.containsKey(AUTHORITIES)) {
-            return Collections.emptyList();
-        }
-        final var authorities = map.get(AUTHORITIES);
-        if (authorities instanceof String) {
-            return AuthorityUtils.commaSeparatedStringToAuthorityList((String) authorities);
-        }
-        if (authorities instanceof Collection) {
-            return AuthorityUtils.commaSeparatedStringToAuthorityList(StringUtils
-                    .collectionToCommaDelimitedString((Collection<?>) authorities));
+    @SuppressWarnings("ConstantConditions")
+    private Collection<GrantedAuthority> extractAuthorities(final Jwt jwt) {
+        final var authorities = new HashSet<>(this.jwtGrantedAuthoritiesConverter.convert(jwt));
+        if (jwt.getClaims().containsKey("authorities")) {
+            authorities.addAll(((Collection<String>) jwt.getClaims().get("authorities")).stream().map(SimpleGrantedAuthority::new).collect(Collectors.toSet()));
```

Comment:
an thing we can do about this nasty cast?

Reply:
I could extend the Spring Jwt class to codify our "authorities" claim as a Collection<String>?  Which essentially just moves the cast...  What do you think?